md/documentation-template.j2

```diff
@@ -0,0 +1,1273 @@
+            - [5.3.3. Fenced Lock Example](#533-fenced-lock-example)
+            - [5.3.4. Map Example](#534-map-example)
+            - [5.3.5. Queue Example](#535-queue-example)
+            - [5.3.6. Set Example](#536-set-example)
+        - [5.4. Multiple Responses to a Single Request](#54-multiple-responses-to-a-single-request)
+        - [5.5. Listeners](#55-listeners)
+        - [5.6. Cluster View Listener](#56-cluster-view-listener)
+        - [5.7. Timeouts and Retry](#57-timeouts-and-retry)
+        - [5.8. Error Codes](#58-error-codes)
+    - [6. Miscellaneous](#6-miscellaneous)
+        - [6.1. Smart Client and Unisocket Client](#61-smart-client-and-unisocket-client)
+        - [6.2. Serialization](#62-serialization)
+        - [6.3. Security](#63-security)
+{# Offset to start listing protocol messages. Should be set to number of manually listed sections + 1 #}
+{% set protocol_message_offset = 7 %}
+    - [{{ protocol_message_offset }}. Protocol Messages](#{{ protocol_message_offset }}-protocol-messages)
+        - [{{ protocol_message_offset }}.1. Custom Data Types Used In The Protocol](#{{ protocol_message_offset }}1-custom-data-types-used-in-the-protocol)
+{% for definition in custom_definitions %}
+{% set custom_types = definition.get('customTypes', None) %}
+{% if custom_types is not none %}
+{% for custom_type in custom_types %}
+            - [{{ protocol_message_offset }}.1.{{ loop.index }}. {{ custom_type.name }}](#{{ protocol_message_offset }}1{{ loop.index }}-{{ custom_type.name|lower }})
+{% endfor %}
+{% endif %}
+{% endfor %}
+{% for service in services %}
+    {% set service_id = service.id + 2 %}
+    - [{{ protocol_message_offset }}.{{ service_id }}. {{ service.name }}](#{{ protocol_message_offset }}{{ service_id }}-{{ service.name|lower }})
+    {% for method in service.methods %}
+        - [{{ protocol_message_offset }}.{{ service_id }}.{{ method.id }}. {{ service.name }}.{{ method.name|capital }}](#{{ protocol_message_offset }}{{ service_id }}{{ method.id }}-{{ service.name|lower }}{{ method.name|lower }})
+    {% endfor %}
+{% endfor %}
+
+{# START of manually written sections #}
+## 1. Introduction
+This document explains the new binary protocol that Hazelcast uses to communicate with the clients.
+This document is not a guide to implement a client that will interact with Hazelcast; rather, it specifies the wire
+data format for the messages exchanged between a client and a Hazelcast member node. Any client that wants to
+communicate with the Hazelcast cluster should obey the data format and communication details explained in this document.
+
+The protocol is designed to be strict enough to ensure standardization in the communication, but flexible enough
+that developers may expand upon the protocol to implement custom features.
+
+General guidelines:
+- This document uses the terms MUST, MUST NOT, MAY, SHOULD, and SHOULD NOT as described in the IETF RFC 2119.
```

Comment:
nit: it might be a good idea to use inline links for RFCs used across this document.